bitcoin/src/crypto/key.rs

```diff
@@ -230,13 +227,52 @@ impl fmt::Display for PublicKey {
     }
 }
 
+/// Error returned while constructing public key from string
+#[derive(Debug, PartialEq)]
+pub enum ParsePublicKeyError {
+    /// Error originated while parsing string
+    Encoding(FromSliceError),
+    /// Hex decoding error.
+    Hex(hex::HexToArrayError),
+    /// `PublicKey` hex should be 66 or 130 digits long.
+    InvalidHexLength(usize),
+}
+
+
+impl From<hex::HexToArrayError> for ParsePublicKeyError {
+    fn from(e: hex::HexToArrayError) -> Self { ParsePublicKeyError::Hex(e) }
+}
+
+impl fmt::Display for ParsePublicKeyError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ParsePublicKeyError::*;
+        match *self {
+            Encoding(ref e) => write_err!(f, "string error"; e),
+            Hex(ref e) => write_err!(f, "hex error"; e),
+            InvalidHexLength(ref got) => write!(f, "pubkey string should be 66 or 130 digits long, got: {}", got),
+        }
+    }
+}
+
+impl From<FromSliceError> for ParsePublicKeyError {
+    fn from(e: FromSliceError) -> ParsePublicKeyError {
+        ParsePublicKeyError::Encoding(e)
+    }
+}
+
 impl FromStr for PublicKey {
-    type Err = Error;
-    fn from_str(s: &str) -> Result<PublicKey, Error> {
+    type Err = ParsePublicKeyError;
+    fn from_str(s: &str) -> Result<PublicKey, ParsePublicKeyError> {
         match s.len() {
-            66 => PublicKey::from_slice(&<[u8; 33]>::from_hex(s)?),
-            130 => PublicKey::from_slice(&<[u8; 65]>::from_hex(s)?),
-            len => Err(Error::InvalidHexLength(len)),
+            66 => {
+                let res = PublicKey::from_slice(&<[u8; 33]>::from_hex(s)?)?;
+                Ok(res)
```

Comment:
In 927e3642415f0d30a47a175ce8eee593c08b9cb4:

```
let res = thing?
Ok(thing)
```

can be replaced everywhere with just `thing` (or possibly `thing.map_err(From::from)`).

Reply:
Thanks!